### ISSUE TITLE
Host header validation, learning from Django

### ISSUE THREAD

**Issue by faassen** [Pylons/webob#174]
I've been reading this document:

http://www.skeletonscribe.net/2013/05/practical-http-host-header-attacks.html

The most interesting bits in this document are the measures the Django project has taken to prevent various attacks.

In Django's `request.py`, there's a `get_host` method. It does the following:
- only respect X-Forwarded-Host if configuration says it can be respected. This relates to pull request #160 which propose to implement this.
- call `split_domain_port` of the host. This in turn checks against a `host_validation_re`. If it's a weird Host header, it'll not return a valid host or port. WebOb doesn't have this validation. Such a validation would be good as the port can be used to smuggle in things we don't want.
- there also seems to be a bit of support for IPv6 I can't detect in the WebOb source.
- after splitting, there's a validation done of the host specifically using `validate_host`. This uses a setting ALLOW_HOSTS to see whether the host is in a whitelist

In addition, I see WebOb there is a `host_port`, but `host_url` does not use it to split `HTTP_HOST`, instead it does it by itself.

Other URL generation code like `application_url` uses `host_url`. This could potentially be used to do a HTML injection attack into URLs generated by WebOb through the HOST header, if at least the URL is put into the template without HTML escaping. But perhaps that's the responsibility of the template language, not WebOb.

Should WebOb implement some of the measures taken by Django? If not, should we at least adjust the WebOb documentation to warn about possible attacks?

**Comment by href** [comment]
Digging this old issue out because I think there would be some good in having Webob help secure all projects that depend on it. Since this should not have anything to do with 'HOST' or 'X-FORWARDED-*' maybe Webob could just assert that the host is valid inside the request class?

This might be a rather implicit thing to do, but it doesn't seem too unreasonable to me to have the attributes of the request class be valid and trusted. This could possibly be implemented on the 'Request' class, leaving the 'BaseRequest' as a fallback for anyone who doesn't want that.

**Comment by renyidong** [comment]
> Developers often resort to the exceedingly untrustworthy HTTP Host header (_SERVER["HTTP_HOST"] in PHP).

That's not correct in WSGI. In WSGI world, the 'HOST' variable is provided by http server. e.g. `$host` from Nginx http://nginx.org/en/docs/http/ngx_http_core_module.html#var_host
So if a "good" http server is used, there shouldn't be such a exploit. 

If Waitress fail to comply the standards, fix Waitress.

**Comment by bertjwregeer** [comment]
@renyidong What if the web server is set to pass through the Host variable because it is listening on an IP address and doesn't have any limits set? Waitress for example just listens on an IP address, and will pass any Host it receives through unmodified. There is no standard saying IP address has to match the hostname, and in some cases (like virtual hosting) it may even be beneficial to allow a variety of Host headers to reach your application.

**Comment by renyidong** [comment]
@bertjwregeer I see your point. Thanks for explain. I should have thought more carefully.

**Comment by corehack** [comment]
So what is the current status on IPv6 support ? Looking at `request.domain`, it doesn't support it at all as it will just split the host incorrectly.
